### PR TITLE
Create Github Action to release winget package

### PR DESCRIPTION
**Pull request by jo-chemla**
Following issue https://github.com/f3d-app/f3d/issues/221 Uses [winget-releaser](https://github.com/vedantmgoyal9/winget-releaser) I suggest a `Classic Github Token` with `public_repo` scope is created by f3d-app org or another maintainer, following [this link](https://github.com/settings/tokens/new), then the Token can be added to the f3d repo as a secret named `WINGET_ACC_TOKEN`. See below, that user also will have to fork the winget-pkgs repository.

Notes:
> You will need to create a *classic* Personal Access Token (PAT) with `public_repo` scope. New fine-grained PATs aren't supported by the action. Review #172 for information.
> Fork [microsoft/winget-pkgs](https://github.com/microsoft/winget-pkgs) under the same account/organization as the project's repository. If you are forking winget-pkgs on a different account (e.g. bot/personal account), you can use the fork-user input to specify the username of the account where the fork is present.

### CODE REVIEW

**Comment by mwestphal** [comment]
While I'm not against this, I fail to undersand why does we (the F3D maintainers and contributors) should be responsible to upload a binary to yet another packaging. Why cant winget packager/logic grab it instead ?

Is there a documentation about the winget workflow and how it is supposed to work

**Comment by mwestphal** [comment]
Also, can/should we upload the nightly too ?

**Comment by jo-chemla** [comment]
The winget-pkgs is a community repository, so anyone can create a manifest for a package - which references the package name, author, installer url etc - for others to install via the `winget install pachage` command. 
Hence why I listed f3d package myself via that PR: https://github.com/microsoft/winget-pkgs/pull/172695

The automation to upload the new installer to winget-pkgs on every release usually happens on the original repo - around 200k PRs, to get an overview of the approximate count of packages hosted in this community repository - so they don't have to watch every release of every package - especially since the packages that use github release are only a portion of the standard winget packages. The github action are the way to do it, but usually involve simply forking the repo, and using the [winget-releaser](https://github.com/vedantmgoyal9/winget-releaser) action to create the new PRs automatically.

File: .github/workflows/winget-release.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    steps:
+      - name: Set winget version env
+        env:
+          TAG_NAME: ${{ inputs.tag_name || github.event.release.tag_name }}
+        run: echo "WINGET_TAG_NAME=$(echo ${TAG_NAME#v})" >> $GITHUB_ENV
+      - name: Submit package to Windows Package Manager Community Repository
+        uses: vedantmgoyal2009/winget-releaser@v2
+        with:
+          identifier: F3d.F3d
+          installers-regex: '^F3D-[\d\.]+-Windows-x86_64.exe\.exe$'
```
Comment (Meakk) [inline .github/workflows/winget-release.yml:26]: `.exe` appears twice in the regex

**Comment by mwestphal** [comment]
Thanks for the explanation.

It happens that our binaries are created by a dedicated release workflow in another repository. I think the action you are adding would be more at home there instead of relying on the `release` trigger.

https://github.com/f3d-app/f3d-superbuild/blob/79c0daeabef09cfa12c2ce17b39ba83c9252d38e/.github/workflows/release.yml#L44

File: .github/workflows/winget-release.yml (hunk 1 above)
Comment (jo-chemla) [reply .github/workflows/winget-release.yml]: woops sorry fixed that typo

**Comment by jo-chemla** [comment]
Thanks for mentioning f3d-superbuild. Usually, the winget-releaser workflow actions are setup on the repo where the release contains the exe or installers, watching for the release. 

I've not seen example of actions where the installer files generated by the build CI are actually used within a subsequent job of the action.

**Comment by mwestphal** [comment]
> Thanks for mentioning f3d-superbuild. Usually, the winget-releaser workflow actions are setup on the repo where the release contains the exe or installers, watching for the release.
> 
> I've not seen example of actions where the installer files generated by the build CI are actually used within a subsequent job of the action.

F3D is different than many other repos as the binary package is not generated manually or from the repo itself but from another repo, the superbuild.

This repository is also responsible for generating our python wheels and publishing them on pypi. It seems logical that the winget package would be uploaded from there too imo, and at the same time.

**Comment by jo-chemla** [comment]
Understood. One thing I forgot to mention, the package manifests references the installer (or portable exe) url, so that every client machine that do run `winget install xyz` request the installer from that url in order to install it. 

So even if the github action is moved outside to f3d-suprebuild, it would reference the installer exe url that resides in https://github.com/f3d-app/f3d/releases/ 
I don't know of existing github actions that allow that kind of behavior.

**Comment by mwestphal** [comment]
> Understood. One thing I forgot to mention, the package manifests references the installer (or portable exe) url, so that every client machine that do run `winget install xyz` request the installer from that url in order to install it.
> 
> So even if the github action is moved outside to f3d-suprebuild, it would reference the installer exe url that resides in https://github.com/f3d-app/f3d/releases/ I don't know of existing github actions that allow that kind of behavior.

I'm not sure to follow, when a user do `winget install f3d`, does it download from https://github.com/f3d-app/f3d/releases/ or from winget servers ?

**Comment by jo-chemla** [comment]
The package manifests are stored on the community repository - and the indexes are usually updated on each machine. So when hte user runs `winget install F3d.F3d`, the manifest which references the metadata and installer url is queried from the index, and the installer is requested through the url listed in the manifest, which for github packages is https://github.com/f3d-app/f3d/releases/

**Comment by mwestphal** [comment]
Ha! So this action does not even uploads the binary package, but just reference a link to it?

**Comment by jo-chemla** [comment]
Yep exactly!

**Comment by mwestphal** [comment]
Then that looks fine to merge imo. Thanks for the explanation.

File: .github/workflows/winget-release.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+name: Submit Winget Package to Windows Package Manager Community Repository
+on:
+  release:
+    types: [released]
+  workflow_dispatch:
+    inputs:
+      tag_name:
+        description: "Specific tag name"
```
Comment (mwestphal) [inline .github/workflows/winget-release.yml:8]: "Specific tag name" ?

Comment (mwestphal) [reply .github/workflows/winget-release.yml:8]: ha right, we are supposed to run this action manually once the release is done, is that correct ?

File: .github/workflows/winget-release.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        required: false
+        type: string
+env:
+  CI: true
```
Comment (mwestphal) [inline .github/workflows/winget-release.yml:12]: why is that needed ?

File: .github/workflows/winget-release.yml (hunk 2 above)
Comment (mwestphal) [reply .github/workflows/winget-release.yml:8]: ok, we can do that but the trigger release above will also trigger it. got it.

File: .github/workflows/winget-release.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set winget version env
+        env:
+          TAG_NAME: ${{ inputs.tag_name || github.event.release.tag_name }}
+        run: echo "WINGET_TAG_NAME=$(echo ${TAG_NAME#v})" >> $GITHUB_ENV
+      - name: Submit package to Windows Package Manager Community Repository
+        uses: vedantmgoyal2009/winget-releaser@v2
+        with:
+          identifier: F3d.F3d
```
Comment (mwestphal) [inline .github/workflows/winget-release.yml:25]: thats a strange name.

F3D is the app
f3d is the project
f3d-app is the org
libf3d is the lib

but F3d is nothing.

File: .github/workflows/winget-release.yml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  CI: true
+jobs:
+  winget-publish:
+    name: Publish winget package
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set winget version env
+        env:
+          TAG_NAME: ${{ inputs.tag_name || github.event.release.tag_name }}
+        run: echo "WINGET_TAG_NAME=$(echo ${TAG_NAME#v})" >> $GITHUB_ENV
```
Comment (mwestphal) [inline .github/workflows/winget-release.yml:21]: what does this do exactly ?

File: .github/workflows/winget-release.yml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+      - name: Submit package to Windows Package Manager Community Repository
+        uses: vedantmgoyal2009/winget-releaser@v2
+        with:
+          identifier: F3d.F3d
+          installers-regex: '^F3D-[\d\.]+-Windows-x86_64\.exe$'
+          version: ${{ env.WINGET_TAG_NAME }}
+          release-tag: ${{ inputs.tag_name || github.event.release.tag_name }}
+          token: ${{ secrets.WINGET_ACC_TOKEN }}
```
Comment (mwestphal) [inline .github/workflows/winget-release.yml:29]: so we need to generate this right ?

File: .github/workflows/winget-release.yml (hunk 3 above)
Comment (jo-chemla) [reply .github/workflows/winget-release.yml:12]: Indeed not needed, I was reusing another github worflow action.

File: .github/workflows/winget-release.yml (hunk 4 above)
Comment (jo-chemla) [reply .github/workflows/winget-release.yml:25]: That's the name of the app that had been submitted to the winget-pkgs repo. What do you think would be a better identifier? `F3D.F3D, f3d.F3D or f3d-app.F3D`? Would need to make a PR to change that identifier before the first run of the action then.

File: .github/workflows/winget-release.yml (hunk 5 above)
Comment (jo-chemla) [reply .github/workflows/winget-release.yml:21]: Stores into an env variable either the name provided by the user firing the github action manually, or the one provided by the automated workflow run on release.

File: .github/workflows/winget-release.yml (hunk 6 above)
Comment (jo-chemla) [reply .github/workflows/winget-release.yml:29]: The github classic token with `public read repo` permission has to be generated by someone on the f3d-app org indeed, and stored as a github secret named `WINGET_ACC_TOKEN` (+ the winget-pkgs repo has to be forked by that same user)

File: .github/workflows/winget-release.yml (hunk 2 above)
Comment (jo-chemla) [reply .github/workflows/winget-release.yml:8]: exact, just understood that! - that action was adapted from another repo. 
By the way, the tag has to be required, updating PR.

File: .github/workflows/winget-release.yml (hunk 4 above)
Comment (mwestphal) [reply .github/workflows/winget-release.yml]: indeed. 

Most definitely `f3d-app.f3d`, that would be coherent with other packages.

File: .github/workflows/winget-release.yml (hunk 5 above)
Comment (mwestphal) [reply .github/workflows/winget-release.yml:19]: my question was specifcally about the `${TAG_NAME#v}` . Why is there a `v` here ?

File: .github/workflows/winget-release.yml (hunk 6 above)
Comment (mwestphal) [reply .github/workflows/winget-release.yml:27]: Why is a fork needed ?

File: .github/workflows/winget-release.yml (hunk 4 above)
Comment (jo-chemla) [reply .github/workflows/winget-release.yml]: Alright, just made a PR to winget-pkgs https://github.com/microsoft/winget-pkgs/pull/173403

File: .github/workflows/winget-release.yml (hunk 5 above)
Comment (jo-chemla) [reply .github/workflows/winget-release.yml:19]: That's a good question, not sure about that. From what I've read, if there is a prefix v1.2.3 in version tag name, then this gets auto stripped by the github action. Seems that this is what this #v does.

**Comment by jo-chemla** [comment]
The PRs to update the package identifier on winget-pkgs have been merged, so now it can be installed via 
```bash
winget source update
winget install f3d-app.f3d
```

File: .github/workflows/winget-release.yml (hunk 6 above)
Comment (mwestphal) [reply .github/workflows/winget-release.yml:27]: noted

File: .github/workflows/winget-release.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+name: Submit Winget Package to Windows Package Manager Community Repository
+on:
+  release:
+    types: [released]
+  workflow_dispatch:
```
Comment (mwestphal) [inline .github/workflows/winget-release.yml:5]: should we generate multiple versions or does the latest version is enough ?

Comment (mwestphal) [reply .github/workflows/winget-release.yml:5]: ping @jo-chemla

Comment (jo-chemla) [reply .github/workflows/winget-release.yml:5]: The new version is probably enough. 

If everything is set-up correctly, every new github release of f3d should therefore automatically create a PR to winget-pkgs - or you can fire a winget-release by executing the action manually and specifying a version tag number. At least this is what is happening now to triplex for which I did the same PR, see [here](https://github.com/try-triplex/triplex/issues/218)

**Comment by mwestphal** [comment]
Ill merge this as soon as I find the time to create the token

**Comment by jo-chemla** [comment]
Hi there, only thing I can see is the regex was maybe wrong, see the suggested edits in [PR](https://github.com/f3d-app/f3d/pull/1934)

**Comment by jo-chemla** [comment]
Ah, just saw the run logs, seems that the github personal access token you provided does not let the CI make a fork, probably the authorizations are not set the correct way? You have that PAT to allow forks. 
There seem to be a bug on github side regarding authorizations for the PAT, see [this issue](https://github.com/russellbanks/Komac/issues/310) in Komac

**Comment by mwestphal** [comment]
So I think I did whats needed, I've added a PAT, I've created a fork, its still not working. Do you have any idea @jo-chemla ?

**Comment by jo-chemla** [comment]
Sorry, from the docs it seems that Komac and winget-releaser@v2 users need a classic github api token rather than the new PAT for the action to run properly. 

> You will need to create a classic Personal Access Token (PAT) with public_repo scope. New fine-grained PATs aren't supported by the action. Review https://github.com/vedantmgoyal9/winget-releaser/issues/172 for information.

The latest execution of the github action is last week Jan 20, after the also failed Jan 19. Seems the error is `Failed to create branch from upstream default branch`, maybe the PAT did not have pull, branch and commit authorization. Anyway, you would need a classic token while the above linked issue is resolved.

**Comment by mwestphal** [comment]
Sorry, I'm removing this action here: https://github.com/f3d-app/f3d/pull/2025

BTW it looks like someone submitted a F3D 3.0 winget package, so all is well :)

https://winstall.app/apps/f3d-app.f3d
